inc/deprecated/DeprecatedClassTrait.php

```diff
@@ -0,0 +1,111 @@
+<?php
+
+namespace WP_Rocket\deprecated;
+
+/**
+ * Trait to use in a deprecated class, or a class containing deprecated methods.
+ */
+trait DeprecatedClassTrait {
+	/**
+	 * Marks a class as deprecated and informs when it has been used.
+	 * Similar to _deprecated_constructor(), but with different strings.
+	 * The current behavior is to trigger a user error if `WP_DEBUG` is true.
+	 *
+	 * @since  3.6
+	 * @author Grégory Viguier
+	 *
+	 * @param string $version     The version of WordPress that deprecated the class.
+	 * @param string $replacement Optional. The method that should have been called. Default null.
+	 */
+	private static function deprecated_class( $version, $replacement = null ) {
+		/**
+		 * Fires when a deprecated class is called.
+		 *
+		 * @since  3.6
+		 * @author Grégory Viguier
+		 *
+		 * @param string $class       The class containing the deprecated constructor.
+		 * @param string $version     The version of WordPress that deprecated the class.
+		 * @param string $replacement Optional. The method that should have been called.
+		 */
+		do_action( 'rocket_deprecated_class_run', self::class, $version, $replacement );
+
+		if ( ! WP_DEBUG ) {
+			return;
+		}
+
+		/**
+		 * Filters whether to trigger an error for deprecated classes.
+		 * `WP_DEBUG` must be true in addition to the filter evaluating to true.
+		 *
+		 * @since  3.6
+		 * @author Grégory Viguier
+		 *
+		 * @param bool $trigger Whether to trigger the error for deprecated classes. Default true.
+		 */
+		if ( ! apply_filters( 'rocket_deprecated_class_trigger_error', true ) ) {
+			return;
+		}
+
+		if ( function_exists( '__' ) ) {
+			if ( ! empty( $replacement ) ) {
+				/**
+				 * With replacement.
+				 */
+				call_user_func(
+					'trigger_error',
+					sprintf(
+						/* translators: 1: PHP class name, 2: version number, 3: replacement class name. */
+						__( 'The called class %1$s is <strong>deprecated</strong> since version %2$s! Use %3$s instead.', 'rocket' ),
+						'<code>' . self::class . '</code>',
+						'<strong>' . $version . '</strong>',
+						'<code>' . $replacement . '</code>'
+					)
+				);
+				return;
+			}
+
+			/**
+			 * Without replacement.
+			 */
+			call_user_func(
+				'trigger_error',
+				sprintf(
+					/* translators: 1: PHP class name, 2: version number. */
+					__( 'The called class %1$s is <strong>deprecated</strong> since version %2$s!', 'rocket' ),
+					'<code>' . self::class . '</code>',
```

Comment:
Change `self::class` to `static::class` for late binding usage.

Reply:
That's the reason I used `self` instead of `static` actually: the class being deprecated is the one using this trait, not necessarily the one extending it imho. And if the class extending the deprecated one must also be deprecated, then use this trait in it. In this case we have the choice.
Or, we could also say that a class that extends a deprecated class is automatically considered deprecated, which I can totally understand.
What do you think @hellofromtonya @crystinutzaa @Tabrisrp?

Reply:
That's an interesting question. IMO, any class extending a deprecated class should be considered deprecated too, or updated to not extend it (extend another one).

Reply:
>Change `self::class` to `static::class` for late binding usage.

>any class extending a deprecated class should be considered deprecated too

OK, let's go with this then? @crystinutzaa Anything against it?

Reply:
Nope. I love the idea that any class which extends the deprecated class will is automatically considered deprecated.

Reply:
OK then. Thank you all :)

Reply:
Done in a37f1d6.